mathjax3-ts/output/chtml/FontData.ts

```diff
@@ -0,0 +1,328 @@
+     *
+     * @param{string} name     The new variant to create
+     * @param{string} inherit  The variant to use if a character is not in this one
+     * @param{string} link     A variant to search before the inherit one (but only
+     *                           its top-level object).
+     */
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
+            this.variant[link].linked.push(variant.chars);
+            variant.chars = Object.create(variant.chars);
+        }
+        this.variant[name] = variant;
+    }
+
+    /*
+     * Create a collection of variants
+     *
+     * @param{string[][]} variants  Array of [name, inherit?, link?] values for
+     *                              the variants to define
+     */
+    public createVariants(variants: string[][]) {
+        for (const variant of variants) {
+            this.createVariant.apply(this, variant);
+        }
+    }
+
+    /*
+     * Defines new character data in a given variant
+     *
+     * @param{srtring} name   The variant for these characters
+     * @param{CharMap} chars  The characters to define
+     */
+    public defineChars(name: string, chars: CharMap) {
+        let variant = this.variant[name];
+        Object.assign(variant.chars, chars);
+        for (const link of variant.linked) {
+            Object.assign(link, chars);
+        }
+    }
+
+    /*
+     * Defines strety delimiters
+     *
+     * @param{DelimiterMap} delims  The delimiters to define
+     */
+    public defineDelimiters(delims: DelimiterMap) {
+        Object.assign(this.delimiters, delims);
+    }
+
+    /*
+     * @param{number} n  The delimiter character number whose data is desired
+     * @return{DelimiterData}  The data for that delimiter (or null)
+     */
+    public getDelimiter(n: number) {
+        return this.delimiters[n];
+    }
+
+    /*
+     * @param{number} n  The delimiter character number whose variant is needed
+     * @param{number} i  The index in the size array of the size whose variant is needed
+     * @return{string}   The variant of the i-th size for delimiter n
+     */
+    public getSizeVariant(n: number, i: number) {
+        if (this.delimiters[n].variants) {
+            i = this.delimiters[n].variants[i];
+        }
+        return this.sizeVariants[i];
+    }
+
+    /*
+     * @param{string} name  The variant whose character data is being querried
+     * @param{number} n     The unicode number for the character to be found
+     * @return{CharData}    The data for the given character (or null)
+     */
+    public getChar(name: string, n: number) {
+        return this.variant[name].chars[n];
```

Comment:
Case that variant[name] does not exist missing.

Reply:
Shouldn't happen, and if it does, it should produce an error (as it will here) so that the inconsistency in the data can be corrected.  This is a function that will be called for every character displayed, so any unnecessary overhead should be avoided.